### DISCUSSION TITLE
Can Quartz.Extensions.DependencyInjectionQuartz.Extensions.DependencyInjection work with .Net Framework?

### DISCUSSION THREAD

**Discussion by stlawrence** [quartznet/quartznet#1649] [Q&A] [answered]
Apologies for the (likely) dumb question.

All the documentation I've found for the Quartz.Extensions.DependencyInjection integration is geared towards the service collection and .Net Core. I see the extension is .Net Standard 2.0 so in theory should be available to use with .Net Framework as well. 

Is there any information on how to configure Quartz.Extensions.DependencyInjection with .Net Framework... ie using the SchedulerBuilder.Create()... method from the tutorial: 
https://www.quartz-scheduler.net/documentation/quartz-3.x/tutorial/using-quartz.html

Thank you for your help!

**Comment by lahma** [comment, marked as answer]
The DI integration intentionally delegates to `SchedulerBuilder` whenever possible so you should get quite the same experience using it. Is there something in particular missing that you are after?

**Reply by stlawrence** [reply]
I appreciate the reply, thanks.

I am building my scheduler, then trying to schedule a job. 

The job has a dependency on a service, but that dependency is not being injected. 
The job succeeds when it has a parameterless constructor but fails otherwise.

How do I tell quartz to use dependency injection when creating the job class? I installed the package but there is no SchedulerBuilder equivalent option for:

// as of 3.3.2 this also injects scoped services (like EF DbContext) without problems
        q.UseMicrosoftDependencyInjectionJobFactory();

Is the problem using SchedulerBuilder? 
Is there an option for setting up dependency injection using StdSchedulerFactory with a config file?

Here is the code to illustrate the use case and issue:

				//Build the Scheduler
				var scheduler = await SchedulerBuilder.Create()
					.UsePersistentStore(x =>
					{
						x.UseProperties = true;
						x.UseClustering();
						x.UseSqlServer(adoProviderOptions =>
						{
							adoProviderOptions.ConnectionString = ConfigurationManager.ConnectionStrings["QuartzDB"].ConnectionString;
							adoProviderOptions.TablePrefix = "[quartz].QRTZ_";
						});
						x.UseJsonSerializer();
					})
					.BuildScheduler();

				//Start Scheduler
				await scheduler.Start();
....

				var systemBackgroundTasksJobkey = new JobKey(nameof(SystemBackgroundTasksJob), Background_Tasks_Group);
				if (!await scheduler.CheckExists(systemBackgroundTasksJobkey))
				{
					var systemBackgroundTasksJob = JobBuilder.Create<SystemBackgroundTasksJob>()
						.WithIdentity(systemBackgroundTasksJobkey)
						.Build();

					await scheduler.ScheduleJob(systemBackgroundTasksJob, systemBackgroundTasksTrigger);
				}

---

	[PersistJobDataAfterExecution]
	[DisallowConcurrentExecution]
	public class SystemBackgroundTasksJob : IJob
	{
		private readonly ISystemBackgroundTasksService _systemBackgroundTasksService;

		//Not working
		//public SystemBackgroundTasksJob(ISystemBackgroundTasksService systemBackgroundTasksService)
		//{
		//	_systemBackgroundTasksService = systemBackgroundTasksService;
		//}

		//ESL workaround until we figure out dependency injection
		public SystemBackgroundTasksJob()
		{
			_systemBackgroundTasksService = DependencyResolver.Current.GetService<ISystemBackgroundTasksService>();
		}

		public async Task Execute(IJobExecutionContext context)
		{
			await _systemBackgroundTasksService.PerformBackgroundTasks();
		}
	}

---

Thanks again.
--Ed

**Reply by lahma** [reply]
So what you need is a job factory, in this case `MicrosoftDependencyInjectionJobFactory`. But the problem is that without the DI infra (which should be available) it cannot easily be built. [Microsoft.Extensions.DependencyInjection is available for full framework though](https://www.nuget.org/packages/Microsoft.Extensions.DependencyInjection/6.0.0#dependencies-body-tab).

So your code would be something similar to this (haven't tested this at all!):

```csharp
var collection = new ServiceCollection();
// probagly need .AddOptions , AddLogging
collection.AddQuartz(x =>
{
    x.UseMicrosoftDependencyInjectionJobFactory();
    // and all the missing bits...
});
var provider = collection.BuildServiceProvider();
var schedulerFactory = provider.GetRequiredService<ISchedulerFactory>();
var scheduler = await schedulerFactory.GetScheduler();
```

But again, it's hard to give advice as I don't know how your system works. Also a possibility to write custom job factory that handles the things, you can check the source here in GitHub what it currently does. There are also providers on NuGet for other DI implementations, like Autofac for example.

I hope this gives you some pointers.

**Reply by stlawrence** [reply]
Thank you. That was the help I needed. 

I changed the code to use the ServiceCollection.AddQuartz method as you suggested and also utilized the Quartz.Unity package you pointed out and was able to use the JobFactory from it via UseJobFactory. There probably needs to be some more fine tuning as I understand this a little better but form now this setup is working.

Thanks again!

				var unityContainer = DependencyResolver.Current.GetService<IUnityContainer>();

				var collection = new ServiceCollection()
				.AddSingleton(unityContainer)
				.AddLogging()
				.AddQuartz(x =>
				{
					x.SchedulerId = "AUTO";
					x.SchedulerName = nameof(Startup);
					x.UseJobFactory<UnityJobFactory>();
					x.UsePersistentStore(storeOptions =>
					{
						storeOptions.UseSqlServer(adoProviderOptions =>
						{
							adoProviderOptions.ConnectionString = ConfigurationManager.ConnectionStrings["QuartzDB"].ConnectionString;
							adoProviderOptions.TablePrefix = "[quartz].QRTZ_";
						});
						storeOptions.UseClustering();
						storeOptions.UseJsonSerializer();
						storeOptions.UseProperties = true;
					});
				});

				var provider = collection.BuildServiceProvider();
				var schedulerFactory = provider.GetRequiredService<ISchedulerFactory>();
				unityContainer.RegisterInstance<ISchedulerFactory>(schedulerFactory);

				var scheduler = await schedulerFactory.GetScheduler();
